User superuser token to gather user information.

thepug opened the pull request:
The storage service needs user information in order to do an auto share.  This change allows the storage service to make a request to get that info.

Supports: https://github.com/open-company/open-company-storage/pull/143

bago2k4 commented on src/oc/auth/api/users.clj:
```diff
@@ -53,17 +53,32 @@
+                 (:super-user (:claims decoded-token)))
+          {:jwtoken decoded-token :user (:claims decoded-token)}
+          false))
       false)))
```
@thepug do we really need to get the token, read it, verify the passphrase etc here?
All that work should be already done by `oc.lib.api.common/open-company-authenticated-resource`, am i missing something here?

thepug replied:
This code doesn't use `oc.lib.api.common/open-company-authenticated-resource`.  `:initialize-context` is where most of that is done but in this case it is overridden and the code here is used.   More reason to put this code into `oc.lib`

bago2k4 replied:
I see because you override `:initialize-context` at line 255. I see you have the JWT set in the same place of a normal request, why you don't add a check for superusers in `oc.lib.api.common`, like `open-company-authenticated-resource` does but it also checks if it's a superuser token.

It will be useful on other services probably.

I'll leave this review to @belucid now, just a suggestion.

thepug replied:
maybe make a tech debt card to move the super user token into lib?

thepug commented:
Also need to fix an issue with normal authenticaton.
```
dev=> 18-08-31 19:28:50 Sunzal.local ERROR [oc.lib.api.common:230] -
                               java.lang.Thread.run              Thread.java:  745
 java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  617
  java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1142
                java.util.concurrent.FutureTask.run          FutureTask.java:  266
java.util.concurrent.Executors$RunnableAdapter.call           Executors.java:  511
                 org.httpkit.server.HttpHandler.run         RingHandler.java:  115
              ring.middleware.reload/wrap-reload/fn               reload.clj:   39
       buddy.auth.middleware/wrap-authentication/fn           middleware.clj:   59
                  ring.middleware.cors/wrap-cors/fn                 cors.clj:   59
                                 clojure.core/apply                 core.clj:  661
                                                ...
                 ring.middleware.cors/allow-origins                 cors.clj:   41
                 ring.middleware.cors/allow-origins                 cors.clj:   48
                 liberator.dev/wrap-trace-header/fn                  dev.clj:  216
                     liberator.dev/wrap-trace-ui/fn                  dev.clj:  212
            liberator.dev/handle-and-add-trace-link                  dev.clj:  194
                        liberator.dev/wrap-trace/fn                  dev.clj:  239
              ring.middleware.params/wrap-params/fn               params.clj:   67
                      raven-clj.ring/wrap-sentry/fn                 ring.clj:   22
                           compojure.core/routes/fn                 core.clj:  192
                                 clojure.core/apply                 core.clj:  659
                                                ...
                             compojure.core/routing                 core.clj:  182
                             compojure.core/routing                 core.clj:  185
                                  clojure.core/some                 core.clj: 2693
                          compojure.core/routing/fn                 core.clj:  185
                           compojure.core/routes/fn                 core.clj:  192
                                 clojure.core/apply                 core.clj:  659
                                                ...
                             compojure.core/routing                 core.clj:  182
                             compojure.core/routing                 core.clj:  185
                                  clojure.core/some                 core.clj: 2693
                          compojure.core/routing/fn                 core.clj:  185
               compojure.core/wrap-route-matches/fn                 core.clj:  146
                  compojure.core/wrap-route-info/fn                 core.clj:  137
            compojure.core/wrap-route-middleware/fn                 core.clj:  128
                    compojure.core/wrap-response/fn                 core.clj:  158
                  compojure.response/eval21424/fn/G             response.clj:    7
                    compojure.response/eval21502/fn             response.clj:   47
                          oc.auth.api.users/user/fn                users.clj:  239
                        liberator.core/run-resource                 core.clj:  595
                  liberator.core/initialize-context                 core.clj:  504
                              liberator.core/decide                 core.clj:   82
                       oc.auth.api.users/user/fn/fn                users.clj:  256
            oc.auth.api.users/allow-superuser-token                users.clj:   59
                                  oc.lib.jwt/decode                  jwt.clj:  154
                              clj-jwt.core/str->jwt                 core.clj:   90
                               clojure.string/split               string.clj:  224
                      java.util.regex.Pattern.split             Pattern.java: 1273
                      java.util.regex.Pattern.split             Pattern.java: 1206
                    java.util.regex.Pattern.matcher             Pattern.java: 1093
                     java.util.regex.Matcher.<init>             Matcher.java:  229
                      java.util.regex.Matcher.reset             Matcher.java:  309
              java.util.regex.Matcher.getTextLength             Matcher.java: 1283
java.lang.NullPointerException:
```

thepug commented:
addressed https://github.com/open-company/open-company-auth/pull/55#issuecomment-418122914 with 41c2fe0d74

belucid commented:
So @thepug this is ready to go again?

thepug commented:
> So @thepug this is ready to go again?


Yeah, I think we can move the token changes to lib later.
